### ISSUE TITLE
gem ruby compatiblity in `gem outdated`

### ISSUE THREAD

**Issue by MSP-Greg** [rubygems/rubygems#2161]
I'm having a problem or would like to suggest a feature.

Using Ruby 2.3.6 & RubyGems 2.7.4, when running `gem outdated`, it reports `did_you_mean (1.0.3 < 1.2.0)`.  the last version of `did_you_mean` that is compatible with 2.3.6 is 1.0.3.  Is this considered an issue?

If I then run `gem update`, the following output:
```
E:\GitHub\ruby-loco>gem update
Updating installed gems
Updating did_you_mean
Fetching: did_you_mean-1.2.0.gem (100%)
ERROR:  Error installing did_you_mean:
        There are no versions of did_you_mean (= 1.2.0) compatible with your Ruby & RubyGems
        did_you_mean requires Ruby version >= 2.5.0dev. The current ruby version is 2.3.0.
Gems updated: did_you_mean
```

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [X] The command line `gem`
  - [ ] Other

Here are my current environment details:

```
RubyGems Environment:
  - RUBYGEMS VERSION: 2.7.4
  - RUBY VERSION: 2.3.6 (2017-12-14 patchlevel 384) [x64-mingw32]
  - INSTALLATION DIRECTORY: E:/GitHub/ruby-loco/pkg/ruby23_64/ruby23_64/lib/ruby/gems/2.3.0
  - USER INSTALLATION DIRECTORY: C:/Users/Greg/.gem/ruby/2.3.0
  - RUBY EXECUTABLE: E:/GitHub/ruby-loco/pkg/ruby23_64/ruby23_64/bin/ruby.exe
  - EXECUTABLE DIRECTORY: E:/GitHub/ruby-loco/pkg/ruby23_64/ruby23_64/bin
  - SPEC CACHE DIRECTORY: C:/Users/Greg/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: C:/ProgramData
  - RUBYGEMS PLATFORMS:
    - ruby
    - x64-mingw32
  - GEM PATHS:
     - E:/GitHub/ruby-loco/pkg/ruby23_64/ruby23_64/lib/ruby/gems/2.3.0
     - C:/Users/Greg/.gem/ruby/2.3.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
     - "install" => "--no-document --env-shebang --platform ruby -- --use-system-libraries"
     - "update" => "--no-document --env-shebang --platform ruby -- --use-system-libraries"
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - E:/GitHub/ruby-loco/pkg/ruby23_64/ruby23_64/bin
     - E:/GitHub/ruby-loco/git/cmd
     - C:\Windows\system32
     - C:\Windows
     - C:\Windows\System32\Wbem
     - C:\Windows\System32\WindowsPowerShell\v1.0
     - C:\ruby26_64\bin
     - E:\msys64
     - C:\Program Files (x86)\GNU\GnuPG\bin
     - C:\Program Files\Git\cmd
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by segiddins** [comment]
This would be nice to improve, but RubyGems as a whole isn't great about suggesting the right gem version based upon compatible ruby/rubygems versions.

**Comment by MSP-Greg** [comment]
I'd like to have a look, as I think that it may become more common for large or default gems to have version restrictions, and also for multiple updates, as Ruby does.

I really haven't torn into the code to see where that logic is, or how complicated it is.  Any suggestions for where to begin looking?

Thanks, Greg

**Comment by segiddins** [comment]
I don't know, sorry

**Comment by MSP-Greg** [comment]
From my recent work, I intent to re-visit this.  Some of the code deals with tuples, some with specs, some with the spec stubs.

Ruby version constraints also affect pre-compiled gems.  I just tried installing a gem on Ruby 2.5.3, and RG choose the pre-compiled gem.  But, it was only built with support for Ruby < 2.5.0.  There was a platform 'ruby' gem available that could be installed...

These issues don't affect me working locally.  I'm interested in them for two reasons, first, new users may find these things confusing.  Secondly, they can add a lot of complexity to CI scripts.

Closing.
